FABulous/FABulous_CLI/FABulous.py

```diff
@@ -0,0 +1,1035 @@
+        self.do_synthesis(str(args.file))
+        self.do_place_and_route(str(json_file_path))
+        self.do_gen_bitStream_binary(str(fasm_file_path))
+
+    @with_category(CMD_SCRIPT)
+    @with_argparser(filePathRequireParser)
+    def do_run_tcl(self, args):
+        """Executes TCL script relative to the project directory, specified by
+        <tcl_scripts>. Uses the 'tk' module to create TCL commands.
+
+        Also logs usage errors and file not found errors.
+
+        Usage:
+            tcl <tcl_scripts>
+
+        Parameters
+        ----------
+        args : str
+            Path to the TCL script.
+        """
+        if not args.file.exists():
+            logger.error(f"Cannot find {args.file}")
+            return
+
+        logger.info(f"Execute TCL script {args.file}")
+        tcl = tk.Tcl()
+        for fun in dir(self.__class__):
+            if fun.startswith("do_"):
+                name = fun.strip("do_")
+                tcl.createcommand(name, getattr(self, fun))
+
+        tcl.evalfile(str(args.file))
+        logger.info("TCL script executed")
+
+
+def main():
+    """Main function to run command line interface for FABulous,
+    sets up argument parsing, initialises required components and handles
+    FABulous CLI execution.
+
+    Also logs terminal output and if .FABulous folder is missing.
+
+    Command line arguments
+    ----------------------
+    Project_dir : str
+        Directory path to project folder.
+    -c, --createProject :  bool
+        Flag to create new project.
+    -csv : str, optional
+        Log all the output from the terminal.
+    -s, --script: str, optional
+        Run FABulous with FABulous script.
+    -log : str, optional
+        Log all the output from the terminal.
+    -w, --writer : <'verilog', 'vhdl'>, optional
+        Set type of HDL code generated. Currently supports .V and .VHDL (Default .V)
+    -md, --metaDataDir : str, optional
+        Set output directory for metadata files, e.g. pip.txt, bel.txt
+    -v, --verbose : bool, optional
+        Show detailed log information including function and line number.
+    -gde, --globalDotEnv : str, optional
+        Set global .env file path. Default is $FAB_ROOT/.env
+    -pde, --projectDotEnv : str, optional
+        Set project .env file path. Default is $FAB_PROJ_DIR/.env
+    """
+    parser = argparse.ArgumentParser(description="The command line interface for FABulous")
+
+    parser.add_argument("project_dir", help="The directory to the project folder")
+
+    parser.add_argument(
+        "-c",
+        "--createProject",
+        default=False,
+        action="store_true",
+        help="Create a new project",
+    )
+
+    parser.add_argument("-csv", default="", nargs=1, help="Log all the output from the terminal")
+
+    parser.add_argument(
+        "-fs",
+        "--FABulousScript",
+        default="",
+        help="Run FABulous with a FABulous script",
+        type=Path,
+    )
+    parser.add_argument(
```

Comment:
The -ts parameter fails for me with our [FABulous.tcl](https://github.com/KelvinChung2000/FABulous/blob/cmd2/FABulous/fabric_files/FABulous_project_template_verilog/FABulous.tcl), but I guess that's not a "real" TCL script. 

![image](https://github.com/user-attachments/assets/729babbc-813e-4bac-944c-63e7420df7ef)

Reply:
The FABulous.tcl is an actual TCL script and a "FABulous script" simultaneously.  I am working on fixing it; the cmd2 argparse is causing a problem with the function execution.